proposals/0297-relax-invalid-nonce.md

```diff
@@ -0,0 +1,81 @@
+---
+simd: '0297'
+title: Relax Invalid Nonced Transactions Constraint
+authors: Tao Zhu (Anza)
+category: Standard
+type: Core
+status: Review
+created: 2025-06-05
+feature:
+---
+
+## Summary
+
+This proposal relaxes the handling of invalid durable nonce transactions during
+block replay. Instead of rejecting the entire block when an invalid nonce is
+encountered, the transaction should be marked as failed, skipped for state
+modifications, but still committed to the block without charging a transaction
+fee.
+
+## Motivation
+
+The current consensus behavior specifies that invalid durable nonce transactions
+should result in block failure during replay. This behavior hinders forward
+compatibility with asynchronous block execution. The goal is to align invalid
+nonce handling with the treatment of other soft transaction failures (e.g.,
+relax fee payer, relax ALT, etc).
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+### Current Behavior
+
+A transaction using a durable nonce fails block replay (causing the entire
+block to be rejected) if any of the following occurs:
+
+- The nonce account is not a statically included account.
+- The nonce account does not exist.
+- The nonce account is not properly initialized.
+- The stored nonce does not match the transaction's recent blockhash.
+- The transaction fails to advance the nonce.
+
+
+### Proposed Change
+
+Update replay logic to treat invalid nonce transactions — those failing the
+checks above — as non-state-modifying and non-fee-charging failures, as follows:
+
+- The transaction is not executed.
+- The transaction is not included in status cache.
+- The transaction’s non-execution CU cost (i.e., transaction's static CUs, plus
+  CUs for loading accounts) still applies to the block limit.
+- The transaction is recorded in the block (marked as failed).
+- No account state is modified, including the nonce account (i.e., nonce is
+  not advanced) and fee payer account (it is not charged with the transaction's
+  fee).
+- The block is not rejected as long as all other transactions replay
+  successfully.
+
+## Alternatives Considered
+
+- Only relax a subset of errors (e.g., missing nonce account but not mismatched
+  blockhash). Adds complexity and partial gains.
+
+## Impact
+
+- Invalid nonced transactions will be included in blocks but will not be charged
+  fees.
+
```

Comment:
We should highlight the impact of downstream effects of this.
Nonced transactions can appear multiple times in the blockchain's history now. This could have a confusing impact on RPCs.

When a user queries a signature, how do I report what slot it's in? Maybe only one that it succeeded in advancing the nonce, if any?

Reply:
Good point about the RPC use case where the same nonced transaction can appear in multiple blocks. Since any invalid instance would be marked as failed, the RPC has flexibility in whether to return the succeeded one or the failed occurrences. In any case, this situation stems from non–state-modifying transactions being included in blocks (as stated), so I don’t think it’s necessary to enumerate specific cases.

Reply:
Well I think it does differ from other non-state-changing transactions because it can have multiple instances (many that change no state, and up to one that can).

afaict the default behavior in our code rn would be to return the first one that appears, though this is probably misleading since that may not be the one that changes state. if rpcs had to check that makes their job harder.

Reply:
Yes, this case is different from other relaxed SIMDs because the same nonced transaction can appear in multiple blocks, as noted in the "Impact". My point was that this impact likely goes beyond just RPC queries, and I didn’t think it was necessary to enumerate every special case. 

However, it might be helpful to explicitly highlight this RPC example. How about to add a sub-bullet like:
```
   - The inclusion of the same nonced transaction in multiple blocks may affect RPC behavior, such as querying a transaction by signature. The exact handling of this is outside the scope of this proposal.
```

Reply:
@ 6b825d62dc807f676e65cd264647fc13fc1e272b